experimental/hip/graph_command_buffer.c

```diff
@@ -0,0 +1,677 @@
+    iree_hal_command_buffer_t* base_command_buffer,
+    iree_hal_buffer_t* source_buffer, iree_device_size_t source_offset,
+    iree_hal_buffer_t* target_buffer, iree_device_size_t target_offset,
+    iree_device_size_t length) {
+  iree_hal_hip_graph_command_buffer_t* command_buffer =
+      iree_hal_hip_graph_command_buffer_cast(base_command_buffer);
+  IREE_TRACE_ZONE_BEGIN(z0);
+
+  const iree_hal_buffer_t* buffers[2] = {source_buffer, target_buffer};
+  IREE_RETURN_AND_END_ZONE_IF_ERROR(
+      z0,
+      iree_hal_resource_set_insert(command_buffer->resource_set, 2, buffers));
+
+  hipDeviceptr_t target_device_buffer = iree_hal_hip_buffer_device_pointer(
+      iree_hal_buffer_allocated_buffer(target_buffer));
+  target_offset += iree_hal_buffer_byte_offset(target_buffer);
+  hipDeviceptr_t source_device_buffer = iree_hal_hip_buffer_device_pointer(
+      iree_hal_buffer_allocated_buffer(source_buffer));
+  source_offset += iree_hal_buffer_byte_offset(source_buffer);
+
+  HIP_MEMCPY3D params = {
+      .srcMemoryType = hipMemoryTypeDevice,
+      .srcDevice = source_device_buffer,
+      .srcXInBytes = source_offset,
+      .dstMemoryType = hipMemoryTypeDevice,
+      .dstDevice = target_device_buffer,
+      .dstXInBytes = target_offset,
+      .WidthInBytes = length,
+      .Height = 1,
+      .Depth = 1,
+  };
+
+  if (command_buffer->graph_node_count >=
+      IREE_HAL_HIP_MAX_CONCURRENT_GRAPH_NODE_COUNT) {
+    return iree_make_status(IREE_STATUS_OUT_OF_RANGE,
+                            "exceeded max concurrent node limit");
+  }
+
+  size_t dependency_count = command_buffer->barrier_node ? 1 : 0;
+  IREE_HIP_RETURN_AND_END_ZONE_IF_ERROR(
+      z0, command_buffer->symbols,
+      hipDrvGraphAddMemcpyNode(
+          &command_buffer->hip_graph_nodes[command_buffer->graph_node_count++],
+          command_buffer->graph, &command_buffer->barrier_node,
+          dependency_count, &params, command_buffer->hip_context),
+      "hipDrvGraphAddMemcpyNode");
+
+  IREE_TRACE_ZONE_END(z0);
+  return iree_ok_status();
+}
+
+static iree_status_t iree_hal_hip_graph_command_buffer_collective(
+    iree_hal_command_buffer_t* base_command_buffer, iree_hal_channel_t* channel,
+    iree_hal_collective_op_t op, uint32_t param,
+    iree_hal_buffer_binding_t send_binding,
+    iree_hal_buffer_binding_t recv_binding, iree_device_size_t element_count) {
+  return iree_status_from_code(IREE_STATUS_UNIMPLEMENTED);
+}
+
+static iree_status_t iree_hal_hip_graph_command_buffer_push_constants(
+    iree_hal_command_buffer_t* base_command_buffer,
+    iree_hal_pipeline_layout_t* pipeline_layout, iree_host_size_t offset,
+    const void* values, iree_host_size_t values_length) {
+  iree_hal_hip_graph_command_buffer_t* command_buffer =
+      iree_hal_hip_graph_command_buffer_cast(base_command_buffer);
+  iree_host_size_t constant_base_index = offset / sizeof(int32_t);
+  for (iree_host_size_t i = 0; i < values_length / sizeof(int32_t); i++) {
```

Comment:
I think we can use a `memcpy` to copy over the whole batch in one go. For example, see how metal side does it.